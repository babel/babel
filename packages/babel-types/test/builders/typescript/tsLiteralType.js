import * as t from "../../../lib/index.js";
import { itBabel7, itBabel8 } from "$repo-utils";

describe("builders", function () {
  describe("typescript", function () {
    describe("tsLiteralType", function () {
      it("accept unary expression", function () {
        expect(
          t.tsLiteralType(t.unaryExpression("-", t.numericLiteral(1))),
        ).toMatchSnapshot();
      });
      itBabel7(
        "accepts unary expression with bigint literal - babel 7",
        function () {
          expect(
            t.tsLiteralType(
              t.unaryExpression("-", t.bigIntLiteral("123456789")),
            ),
          ).toMatchInlineSnapshot(`
            Object {
              "literal": Object {
                "argument": Object {
                  "type": "BigIntLiteral",
                  "value": "123456789",
                },
                "operator": "-",
                "prefix": true,
                "type": "UnaryExpression",
              },
              "type": "TSLiteralType",
            }
          `);
        },
      );
      itBabel8("accepts unary expression with bigint literal", function () {
        expect(
          t.tsLiteralType(
            t.unaryExpression("-", t.bigIntLiteral(BigInt("123456789"))),
          ),
        ).toMatchInlineSnapshot(`
          Object {
            "literal": Object {
              "argument": Object {
                "type": "BigIntLiteral",
                "value": 123456789n,
              },
              "operator": "-",
              "prefix": true,
              "type": "UnaryExpression",
            },
            "type": "TSLiteralType",
          }
        `);
      });
      it("throws with non-numeric argument", function () {
        expect(() => {
          t.tsLiteralType(t.unaryExpression("-", t.stringLiteral(1)));
        }).toThrow("Property value expected type of string but got number");
        expect(() => {
          t.tsLiteralType(t.unaryExpression("-", t.objectExpression([])));
        }).toThrow(
          'Property argument of UnaryExpression expected node to be of a type ["NumericLiteral","BigIntLiteral"] but instead got "ObjectExpression"',
        );
      });
    });
    it("throws with bad operator", function () {
      expect(() => {
        t.tsLiteralType(t.unaryExpression("+", t.numericLiteral(1)));
      }).toThrow(
        'Property operator expected value to be one of ["-"] but got "+"',
      );

      // should check operator first since it appears first
      expect(() => {
        t.tsLiteralType(t.unaryExpression("+", t.objectExpression([])));
      }).toThrow(
        'Property operator expected value to be one of ["-"] but got "+"',
      );

      expect(() => {
        t.tsLiteralType(t.unaryExpression("~", t.numericLiteral(1)));
      }).toThrow(
        'Property operator expected value to be one of ["-"] but got "~"',
      );

      expect(() => {
        t.tsLiteralType(t.unaryExpression("void", t.numericLiteral(1)));
      }).toThrow(
        'Property operator expected value to be one of ["-"] but got "void"',
      );
    });
  });
});
