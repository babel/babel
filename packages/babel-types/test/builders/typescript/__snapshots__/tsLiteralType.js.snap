// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builders typescript tsLiteralType accept unary expression 1`] = `
Object {
  "literal": Object {
    "argument": Object {
      "type": "NumericLiteral",
      "value": 1,
    },
    "operator": "-",
    "prefix": true,
    "type": "UnaryExpression",
  },
  "type": "TSLiteralType",
}
`;
